### PR TITLE
Change wait time between batch submit responsively

### PR DESCRIPTION
**Pull request by Caleb-Hill**
Transact waits the full time between the end of one batch submit and the
next regardless of how long the submit takes. This can make the target
submission rate and the actual submission rate be very different.
This change alters the time between submissions based on how long the
current submission took and a "best guess" for how long they usually
take.
Ideally this will more tightly tie the target and actual submission rate.

Signed-off-by: Caleb Hill <hill@bitwise.io>

### CODE REVIEW

File: libtransact/src/workload/runner.rs [hunk 1]
```diff
@@ -33,6 +33,7 @@ use super::error::WorkloadRunnerError;
 use super::BatchWorkload;
 
 pub const DEFAULT_LOG_TIME_SECS: u32 = 30; // time in seconds
+const SMOOTHING_FACTOR: f32 = 0.5; //how much to smooth transaction time estimates
```
Comment (vaporos) [inline libtransact/src/workload/runner.rs:36]: Add a space between // and "how"

File: libtransact/src/workload/runner.rs [hunk 2]
```diff
@@ -338,7 +341,18 @@ impl WorkerBuilder {
 
                                 // get next target, round robin
                                 next_target = (next_target + 1) % targets.len();
-                                thread::sleep(time_to_wait);
+                                let diff = time::Instant::now() - submission_start;
+                                let submission_time = match submission_avg {
+                                    Some(val) => {
```
Comment (vaporos) [inline libtransact/src/workload/runner.rs:346]: This looks like it reduces to avg=(diff+avg)/2; suggest we simplify the implementation here until we need a more complex approach.

**Review by vaporos — CHANGES_REQUESTED** [review]
There is a typo in the commit "bath" instead of "batch" and a empty line missing between paragraphs.

File: libtransact/src/workload/runner.rs (hunk 2 above)
Comment (Caleb-Hill) [reply libtransact/src/workload/runner.rs]: Simplified to just that expression

File: libtransact/src/workload/runner.rs [hunk 3]
```diff
@@ -338,7 +340,15 @@ impl WorkerBuilder {
...
                                 next_target = (next_target + 1) % targets.len();
-                                thread::sleep(time_to_wait);
+                                let diff = time::Instant::now() - submission_start;
+                                let submission_time = match submission_avg {
+                                    Some(val) => (diff + val) / 2,
+                                    None => diff,
+                                };
+                                submission_avg = Some(submission_time);
+
+                                thread::sleep(time_to_wait - submission_time);
```
Comment (agunde406) [inline libtransact/src/workload/runner.rs:350]: I see this throwing the following panic `thread 'Smallbank-Workload-0' panicked at 'overflow when subtracting durations', library/core/src/time.rs:890:31`

Comment (agunde406) [reply libtransact/src/workload/runner.rs:350]: this happens when the TooManyRequest loop has happened.

**Review by agunde406 — CHANGES_REQUESTED** [review]

File: libtransact/src/workload/runner.rs (hunk 3 above)
Comment (Caleb-Hill) [reply libtransact/src/workload/runner.rs]: Yup, should have seen that coming. Changed it to a saturating subtraction so the floor is zero always.

File: libtransact/src/workload/runner.rs [hunk 4]
```diff
@@ -554,7 +566,15 @@ pub fn submit_batches_from_source(
...
         next_target = (next_target + 1) % targets.len();
-        thread::sleep(time_to_wait);
+        let diff = time::Instant::now() - submission_start;
+        let submission_time = match submission_avg {
+            Some(val) => (diff + val) / 2,
+            None => diff,
+        };
+        submission_avg = Some(submission_time);
+
+        thread::sleep(time_to_wait - submission_time);
```
Comment (agunde406) [inline libtransact/src/workload/runner.rs:576]: This should also be updated to use time_to_wait.saturating_sub

**Review by agunde406 — CHANGES_REQUESTED** [review]

File: libtransact/src/workload/runner.rs (hunk 4 above)
Comment (Caleb-Hill) [reply libtransact/src/workload/runner.rs]: Yeah, pushed it and then saw the issue. Just pushed yet another fix. Thanks

**Review by agunde406 — APPROVED** [review]

**Review by vaporos — APPROVED** [review]

**Pull request merged**